promremote/client.go

```diff
@@ -0,0 +1,178 @@
+// THE SOFTWARE.
+
+package promremote
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/m3db/m3/src/query/ts"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/prometheus/prometheus/prompb"
+)
+
+const (
+	// DefaultRemoteWrite is the default Prom remote write endpoint in m3coordinator.
+	DefaultRemoteWrite = "http://localhost:7201/api/v1/prom/remote/write"
+
+	defaulHTTPClientTimeout = time.Second * 30
+)
+
+// Tag are the metric tags.
+type Tag struct {
+	Name  string `json:"name"`
+	Value string `json:"value"`
+}
+
+// Timeseries are made of tags and a datapoint.
+type Timeseries struct {
+	Tags      []Tag
+	Datapoint ts.Datapoint
+}
+
+type client struct {
+	writeURL   string
+	httpClient *http.Client
+}
+
+type clientOptions struct {
+	remoteWriteURL    string
+	httpClientTimeout time.Duration
+}
+
+// M3Client is used to write timeseries data to m3coordinator.
+type M3Client interface {
+	// Write writes the Prom proto WriteRequest to the specified endpoint.
+	Write(*prompb.WriteRequest) error
+}
+
+// ClientOptions defines available methods.
+type ClientOptions interface {
+	// SetWriteURL sets the URL which the client uses to write to m3coordinator.
+	SetWriteURL(string) ClientOptions
+
+	// WriteURL returns the URL which the client uses to write to m3coordinator.
+	WriteURL() string
+
+	// SetHTTPClientTimeout sets the timeout for the client.
+	SetHTTPClientTimeout(time.Duration) ClientOptions
+
+	//HTTPClientTimeout returns the timeout that is set for the client.
+	HTTPClientTimeout() time.Duration
+}
+
+// NewClientOpts returns a default clientOptions struct.
+func NewClientOpts() ClientOptions {
+	return &clientOptions{
+		remoteWriteURL:    DefaultRemoteWrite,
+		httpClientTimeout: defaulHTTPClientTimeout,
+	}
+}
+
+// NewClient creates a new remote write coordinator client.
+func NewClient(opts ClientOptions) M3Client {
+	return &client{
+		writeURL: opts.WriteURL(),
+		httpClient: &http.Client{
+			Timeout: opts.HTTPClientTimeout(),
+		},
+	}
+}
+
+func (o *clientOptions) SetWriteURL(val string) ClientOptions {
+	opts := *o
+	opts.remoteWriteURL = val
+	return &opts
+}
+
+func (o *clientOptions) WriteURL() string {
+	return o.remoteWriteURL
+}
+
+func (o *clientOptions) SetHTTPClientTimeout(val time.Duration) ClientOptions {
+	opts := *o
+	opts.httpClientTimeout = val
+	return &opts
+}
+
+func (o *clientOptions) HTTPClientTimeout() time.Duration {
+	return o.httpClientTimeout
+}
+
+func (c *client) Write(promWR *prompb.WriteRequest) error {
+	data, err := proto.Marshal(promWR)
+	if err != nil {
+		return fmt.Errorf("unable to marshal protobuf: %v", err)
+	}
+
+	encoded := snappy.Encode(nil, data)
+
+	body := bytes.NewReader(encoded)
+	req, err := http.NewRequest("POST", c.writeURL, body)
+	if err != nil {
+		return err
+	}
+
+	req.Header.Set("Content-Type", "application/x-protobuf")
+	req.Header.Set("Content-Encoding", "snappy")
```

Comment:
Probably would add the other 2 headers that prom remote write sends:
```
	req.Header.Add("Content-Encoding", "snappy")
	req.Header.Set("Content-Type", "application/x-protobuf")
	req.Header.Set("User-Agent", userAgent)
	req.Header.Set("X-Prometheus-Remote-Write-Version", "0.1.0")
```

Maybe add a default user agent with `"promremote-go/1.0.0"` and allow it to be overridden if a non-empty `UserAgent string` on `Config` is set.